Add a grid-wise memory fence

fwyzard opened the pull request:
A grid-wise memory fence is a no-op in the CPU serial case, running with a single thread; and is implemented as a per-device fence in the CUDA/HIP, OpenMP 2, OpenMP 5 and SYCL backends, that do not have a native per-grid fence.

As usual there are no fences for the OpenAcc backend.

Documentation and tests have been updated accordingly.

fwyzard committed:
Add a grid-wise memory fence

A grid-wise memory fence is a no-op in the CPU serial case, running with
a single thread; and is implemented as a per-device fence in the CUDA/HIP,
OpenMP 2, and OpenMP 5 backends, that do not have a native per-grid fence.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard commented:
As usual there are no fences for the OpenAcc backend.

fwyzard commented:
I did not try to implement a grid-wise fence for the experimental SYCL backend.

fwyzard commented:
@tonydp03 @j-stephan FYI

j-stephan commented:
> I did not try to implement a grid-wise fence for the experimental SYCL backend.

There is actually no need for that. The SYCL back-end has a special case for the block-level fence and will resort to the device-wide fence otherwise.

j-stephan reviewed (changes requested):
Looks solid to me. Could you update the copyright headers, too?

fwyzard commented:
> Could you update the copyright headers, too?

Done.

fwyzard commented:
> The SYCL back-end has a special case for the block-level fence and will resort to the device-wide fence otherwise.

I see, thanks !

fwyzard commented:
As expected, using this new  level of fences our application achieved the same performance as using the original implementation: https://github.com/cms-patatrack/pixeltrack-standalone/pull/320#issuecomment-1065378932 .

bernhardmgruber reviewed (changes requested):
Thank you for this PR! The implementaiton LGTM.

I would also like to see this new feature being called in one of the unit tests. And the cheatsheet mentions memory fences, so the new grid level option should be mentioned there as well.

fwyzard committed:
Update documentation for memory fences

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard commented:
> I would also like to see this new feature being called in one of the unit tests.

The current test based on  `DeviceFenceTestKernel` is actually testing for grid-wise fences, since it involves a single kernel.


> And the cheatsheet mentions memory fences, so the new grid level option should be mentioned there as well.

OK, I've added it along with the block and device cases.

fwyzard committed:
Rewrite the memory fence tests

Update the checks to reject only the case that is prevented by the fence.

Make sure the block-wise test runs with two threads, or at two elements if
the accelerator does not support multiple threads per block.

Make sure the existing device-wise fence test runs with two blocks, and
repurpose it for testing grid-wise fences.

Add a dedicated test for device-wise fences, using two concurrent kernels
running on the same device.

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard commented:
> I would also like to see this new feature being called in one of the unit tests.

Done: I've repurposed the original device-wise fence test using a single kernel as a test for the grid-wise fences, ~~and added a dedicated test using two potentially concurrent kernels for the device-wise fences.~~

fwyzard commented:
OK, the grid-wise fence works.
I'll try to see what was the problem with the device-wise test I wrote in a separate PR.

fwyzard commented:
See #1642 for the new test.

j-stephan commented:
~~@bernhardmgruber Please don't merge before #1642 has passed the CI.~~

psychocoderHPC commented on docs/source/basic/library.rst:
```diff
@@ -161,9 +161,9 @@ guarantees the following **for the local thread**  and regardless of global or s
-threads in the same block or the whole device.
+Memory fences can be issued on the block level (``alpaka::memory_scope::Block``), grid level
+(``alpaka::memory_scope::Grid``) and the device level (``alpaka::memory_scope::Device``).
+Depending on the memory scope, the *StoreStore* order will be visible to other threads in the same block or the whole device.
```
Currently, the new option `Grid` is not explained in the documentation string.

should be `Depending on the memory scope, the *StoreStore* order will be visible to other threads in the same block, other blocks in a grid or the whole device (grids between grids).`

fwyzard replied:
Done.

fwyzard commented:
> @bernhardmgruber Please don't merge before #1642 has passed the CI.

Why not ?
The new code I added is being tested successfully.

j-stephan commented:
> Why not ?

My bad, I wrote this before looking at #1642. I assumed you would test the new feature there.

fwyzard commented:
By the way, is the old test actually testing the grid-wise fence ?
Isn't `fixture` running with a single block, single thread ?

fwyzard commented:
I've come to the conclusion that both the test and the check are wrong :-/

The comments in the code with the implementations of the fences say
> The following are all valid outcomes:
>    `a == 1 && b == 2`
>    `a == 10 && b == 2`
>    `a == 10 && b == 20`

So the check should be `! (a == 1 && b == 20)`, or `a != 1 || b != 20`.

Instead the check requires `a != 1 && b == 20`.

so far the tests have all passed because they run a single kernel with a single block and a single thread -- so they effectively run serially the whole thing:
```c++
vars[0] = 10;
vars[1] = 20;
auto const b = vars[1];
auto const a = vars[0];
```
which is equivalent to setting `a = 10` and `b = 20`.

fwyzard committed:
Work around a buggy -Wmissing-variable-declarations warning in clang 5 and 6

Signed-off-by: Andrea Bocci <andrea.bocci@cern.ch>

fwyzard commented:
And I've added the usual workaround for the buggy `-Wmissing-variable-declarations` warning in clang 5 and 6.

bernhardmgruber commented on test/unit/mem/fence/src/FenceTest.cpp:
```diff
@@ -14,7 +14,86 @@
 
-class DeviceFenceTestKernel
+// Trait to detect whether an accelerator supports or not multiple threads per block
+template<typename TAcc>
+struct IsSingleThreaded : public std::false_type
+{
+};
```
I would just make this a variable template directly and then specialize it later:
```suggestion
template<typename TAcc>
inline constexpr bool isSingleThreaded = false;
```

fwyzard replied:
I see... I would prefer to make this change globally, also in the other places that use template variables.

fwyzard commented:
Looks like the failure is due to a runner getting stuck ?

> ERROR: Job failed: execution took longer than 3h0m0s seconds

fwyzard commented:
> Looks like the failure is due to a runner getting stuck ?

Should I just re-trigger the tests ?
Any way to do it other than force-pushing to the branch ?

j-stephan commented:
Sorry for the late response, we just had our alpaka VC. This looks like the runner GPU crashed. We informed our IT department and are now waiting for the node to be rebooted. Afterwards we can restart the job, there is no need to do a force-push.

psychocoderHPC commented:
> > Looks like the failure is due to a runner getting stuck ?
> 
> Should I just re-trigger the tests ? Any way to do it other than force-pushing to the branch ?

The runner is back in action and I restarted the failing test.

fwyzard commented:
:+1:
